bot/seasons/season.py

```diff
@@ -52,82 +59,289 @@ def get_season(bot, season_name: str = None, date: datetime.date = None):
+            return await resp.read()
+
+    async def apply_username(self, *, debug: bool = False) -> Union[bool, None]:
+        """
+        Applies the username for the current season. Only changes nickname if
+        `bool` is False, otherwise only changes the nickname.
+
+        Returns True if it successfully changed the username.
+        Returns False if it failed to change the username, falling back to nick.
+        Returns None if `debug` was True and username change wasn't attempted.
         """
 
-        guild = self.bot.get_guild(Client.guild)
+        guild = bot.get_guild(Client.guild)
+        result = None
 
         # Change only nickname if in debug mode due to ratelimits for user edits
-        if Client.debug:
+        if debug:
             if guild.me.display_name != self.bot_name:
                 log.debug(f"Changing nickname to {self.bot_name}")
                 await guild.me.edit(nick=self.bot_name)
+
         else:
-            if self.bot.user.name != self.bot_name:
+            if bot.user.name != self.bot_name:
                 # attempt to change user details
                 log.debug(f"Changing username to {self.bot_name}")
-                await self.bot.user.edit(name=self.bot_name, avatar=self.bot_avatar)
+                try:
+                    await bot.user.edit(username=self.bot_name)
+                except discord.HTTPException:
+                    pass
 
                 # fallback on nickname if failed due to ratelimit
-                if self.bot.user.name != self.bot_name:
-                    log.info(f"User details failed to change: Changing nickname to {self.bot_name}")
+                if bot.user.name != self.bot_name:
+                    log.warning(f"Username failed to change: Changing nickname to {self.bot_name}")
                     await guild.me.edit(nick=self.bot_name)
+                    result = False
+                else:
+                    result = True
 
             # remove nickname if an old one exists
             if guild.me.nick and guild.me.nick != self.bot_name:
                 log.debug(f"Clearing old nickname of {guild.me.nick}")
                 await guild.me.edit(nick=None)
 
+            return result
+
+    async def apply_avatar(self) -> bool:
+        """
+        Applies the avatar for the current season. Returns if it was successful.
+        """
+
+        # track old avatar hash for later comparison
+        old_avatar = bot.user.avatar
+
+        # attempt the change
+        log.debug(f"Changing avatar to {self.icon}")
+        icon = await self.get_icon()
+        try:
+            async with async_timeout.timeout(5):
+                await bot.user.edit(avatar=icon)
+        except (discord.HTTPException, asyncio.TimeoutError):
+            pass
+
+        if bot.user.avatar != old_avatar:
+            log.debug(f"Avatar changed to {self.icon}")
+            return True
+
+        log.warning(f"Changing avatar failed: {self.icon}")
```

Comment:
Why wouldn't we directly check the exceptions, though?

Reply:
Because failure can occur without an exception.